### PR TITLE
[#73523] Implement `WorkPackage` semantic ID resolution system (storage-optimized)

### PR DESCRIPTION
**Pull request by thykel**
# Ticket
https://community.openproject.org/projects/communicator-stream/work_packages/73436
https://community.openproject.org/projects/communicator-stream/work_packages/73523

## What are you trying to accomplish?

Add support for semantic WP identifiers in the format of `{project identifier}-{sequence number}` (e.g. "PROJ-42"). 

The semantic WP identifiers have to be stable across time, i.e. even if the identifier changes multiple times (due to project rename or WP move), all old identifiers must still resolve to the same WP.

Additionally, we need to support "ghost identifiers", which are identifiers that have technically never been generated by the system, but we should still somehow support them to allow flexible aliases. Example:

1. Project `PROJ` contains WP `PROJ-1`
2. `PROJ` gets renamed to `PROJ_NEW`
3. `PROJ_NEW` has a new WP added 
4. `PROJ-2` needs to resolve to `PROJ_NEW-2`

## What approach did you choose and why?

* Add `Project.wp_sequence_counter` to track per-project WP sequences
* Add `WorkPackage.sequence_number` to provide numeric representation of the sequence part of the ID
* Add `WorkPackage.semantic_id` to reflect the current semantic ID of the work package
* Add `WorkPackageSemanticAlias` to track the historical handles for each WP
  * The table contains the following two kinds of items:
    * All pre-project-rename identifiers across history (i.e. for `GREEN`->`BLUE`->`RED`, the table contains pointers for every WP handle of `GREEN` and `BLUE`, e.g. `GREEN-1` and `BLUE-1`)
    * All pre-package-move identifiers across history (i.e. if a WP was moved from `GREEN-1` to `BLUE-12` and then `RED-299`, the table contains pointers for handles `GREEN-1` and `BLUE-12`)

The flow consists of the following steps, firing off a single query for the happy path, and 5 queries for the absolute worst case scenario.

1. **Resolve valid WP identifiers:** Indexed string lookup in `WorkPackage`.
2. **Resolve moved WPs & pre-rename identifiers:** Indexed string lookup in `WorkPackageSemanticAlias`
3. **Resolve pre-rename ghost identifiers (`{pre-rename project ID}-{post-rename sequence ID}`):** `Project` record lookup using `friendly_id` (which supports historical lookups) +`WorkPackage` lookup using sequence number
4. **Resolve pre-rename'n'move ghost identifiers (`{pre-rename project ID}-{post-rename moved WP sequence ID}`):** `WorkPackageSemanticAlias` lookup using a "refreshed" semantic identifier (using the current project identifier. e.g. NEWPROJ-4 instead of input OLDPROJ-4)
5. `404`.

**tl;dr version:** We first try to look up the WP table + alias table using the _given_ raw semantic ID, and then we validate + "refresh" the identifier using `friendly_id` and look up the WP + alias table again.

**Tradeoffs:**
  - The alias table hoards all historical identifiers after renames, which can sometimes add up -- e.g. if a project with 10k WPs gets its identifier changed, the alias table receives 10k new records.
    - This is deemed acceptable as we could use that O(1) access and each row is quite tiny (just the handle + integer fkey). However, if we decide that saving DB space is more important than lookup speed, we can also very easily _not_ write into the table upon rename -- which will mean that lookup step 2 will be a miss and the WP will be still correctly located by the manual fallback in step 3.
    - Something that's non-negotiable is that the alias table needs to track _renamed_ packages. (These will scale a lot better than the rename example above.)
  - The worst case scenario currently comes up to 5 queries. We _are_ able to cut this down to just 4 queries by squashing steps 3 and 4 together, but that requires us to further denormalize the database by also storing & maintaining all _current_ semantic IDs in the alias table. This would now affect the WP create operation and introduce potential for state drift. The current option honestly still seems better, as the "expensive" edge case is quite obscure and the performance hit is minor.

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

File: app/models/work_package/identifier.rb [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module WorkPackage::Identifier
+  extend ActiveSupport::Concern
+
+  SEMANTIC_PATTERN = /\A([A-Za-z][A-Za-z0-9_]*)-(\d+)\z/
+
+  included do
+    has_many :semantic_ids,
```
Comment (thykel) [inline app/models/work_package/identifier.rb]: Better naming might be `:semantic_id`/`:semantic_aliases`

File: app/controllers/work_packages_controller.rb [hunk 2]
```diff
@@ -243,7 +243,7 @@ def project
   def work_package
     return @work_package if defined?(@work_package)
 
-    @work_package = WorkPackage.visible(current_user).find_by(id: params[:id])
+    @work_package = WorkPackage.find_by_identifier(params[:id], user: current_user)
```
Comment (thykel) [inline app/controllers/work_packages_controller.rb]: Note: I think we can continue working with that `visible` scope, this refactoring bit looks hallucinated.

File: app/models/work_package/identifier.rb [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      user ? visible(user) : all
+    end
+  end
+
+  private
+
+  def register_semantic_id
+    seq = project.with_lock { project.increment!(:wp_sequence_counter).wp_sequence_counter }
+    update_columns(sequence_number: seq)
+    WorkPackageSemanticId.create!(identifier: "#{project.identifier}-#{seq}", work_package_id: id, current: true)
```
Comment (thykel) [inline app/models/work_package/identifier.rb]: todo: should be a transaction

**Comment by judithroth** [comment]
Thanks for creating this POC!

Before we dive into details here (I see a lot of details that we should change before this gets final, but this is only a POC, so I try to ignore the details).

Here are some main points I think we should address right now:
1) When the instance is in semantic ID mode, almost every read of a work package will involve also its semantic identifier (e.g. to show it in the view or API or whatever). So if we keep the full semantic ID only in the join table, we will have to add a lot of `includes` or get a lot of n+1 queries. That's why I believe - independent of how we store history - that we need a column `identifier` on the `work_packages` table with the full identifier (e.g. `PROJ-11`), not only the sequence number.
2) If we do that, we can also consider to first search in the `work_packages` table for a match and only if that returns `nil` fall back to searching in the registry. This makes the primary use case (finding by a current identifier) faster (no join needed), but would make the secondary use case (finding by an historic identifier) slower (since it's one more query). 
3) When we need to replicate entries in the `work_package_semantic_ids` table because of a project identifer change, we can not rely on the `current` flag, since we also need to replicate entries that belong to work packages that were moved out of the project and into a different project (and therefore no longer have the `current` flag, see `WorkPackageSemanticId.register_move`). So we always need to select by the prefix of the previous project identifier. I think if we do suggestion no. 1 then we can drop the `current` flag completely.
4) I think for the `BackfillService` we should already use the counter column approach that we discussed over the `max+1` approach 

@thykel and @akabiru - do you agree?

File: app/models/work_package/semantic_identifier.rb [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    after_create :register_semantic_id, if: -> { Setting::WorkPackageIdentifier.alphanumeric? }
+  end
+
+  class_methods do
+    # Resolves any identifier form to a WorkPackage, applying visibility when user is given.
+    #   - Numeric string ("12345")    → find by primary key
+    #   - Semantic string ("PROJ-42") → registry lookup, then computed fallback
+    #
+    # Returns nil on miss.
+    def find_by_identifier(identifier)
```
Comment (thykel) [inline app/models/work_package/semantic_identifier.rb:54]: Obviously let's not focus on the actual finder integration for now -- we'll normalize that via a follow-up PR.

**Comment by thykel** [comment]
> When the instance is in semantic ID mode, almost every read of a work package will involve also its semantic identifier (e.g. to show it in the view or API or whatever). So if we keep the full semantic ID only in the join table, we will have to add a lot of `includes` or get a lot of n+1 queries. That's why I believe - independent of how we store history - that we need a column `identifier` on the `work_packages` table with the full identifier (e.g. `PROJ-11`), not only the sequence number.

Done! However, this also implies we may not actually need to maintain the current identifiers in the registry/alias table. 

I tweaked the design to have the alias table only store _pre-rename identifiers_ and _WP move mappings_. The PR description goes into this in detail. I think the result is a pretty interesting trade-off.



> I think for the `BackfillService` we should already use the counter column approach that we discussed over the `max+1` approach

Let's discuss that further on a call over the [follow-up PR](https://github.com/opf/openproject/pull/22566).

File: app/models/work_package/semantic_identifier.rb [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    # Covers "ghost" identifiers arising from project renames and move chains:
+    #   - "PROJ-2" after PROJ → PROJ_NEW (new WP got PROJ_NEW-2, no PROJ-2 row).
+    #   - "PROJ-5" after PROJ → PROJ_NEW then WP moved to OTHER ("PROJ_NEW-5" alias
+    #     exists; reconstruct from current project name and look up in alias table).
+    def find_by_computed_fallback(identifier)
+      prefix, seq = parse_semantic_identifier(identifier)
+      return nil unless prefix && seq
+
+      project = resolve_project_by_prefix(prefix)
+      return nil unless project
```
Comment (thykel) [inline app/models/work_package/semantic_identifier.rb:162]: TODO: At this point we can short-circuit the whole thing if the given project identifier was actually the current one. Saves us two extra queries. 🥳

**Comment by thykel** [comment]
superseded by https://github.com/opf/openproject/pull/22581
